torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -309,6 +357,20 @@ def foreach_reduce(
         reduce_scatter_group, all_reduce_group, reduce_dtype
     )
     world_size = reduce_scatter_group.size()
+    for i, (fsdp_param, unsharded_grad) in enumerate(zip(fsdp_params, unsharded_grads)):
+        shard_dim = fsdp_param.fsdp_placement.dim
+        if shard_dim != 0:
+            if unsharded_grad.size(shard_dim) % world_size != 0:
```

Comment:
For now, I could not come up with a way to avoid this extra copy for unsharded gradients that require padding when we are not sharding on dim-0.

cc: @yifuwang in case you have any ideas
(more generally, if you look through the changes, there are a few places where we did 'tricks' like `torch.as_strided` with contiguous strides to reduce CPU overhead, but those fell apart once sharding on non-zero dim)